Web Extensions Declarative Net Request WIP

jdatapple opened the pull request:
#### Summary
A work in progress branch for adding Declarative Net Request API and manifest support for Web Extensions.

Elchi3 commented:
Looks alright to me but would appreciate if @rebloor, @Rob--W or @rpl would take a look since I don't know WebExtensions all that well :)

rebloor commented:
It's a little difficult to comment without some context on this. Also, have you run `npm test` and confirmed that it has reported no errors.

rpl commented:
@Rob--W would you mind to provide some feedback on this? (in its current form it is mainly a skeleton of the structure meant to track DeclarativeNetRequest implementation across browser vendors, while the `version` values associated to each browser are not yet populated).

queengooborg commented on webextensions/api/declarativeNetRequest.json:
```diff
@@ -0,0 +1,316 @@
+        },
+        "onRuleMatchedDebug_event": {
+          "__compat": {
+            "mdn_url": "https://developer.mozilla.org/docs/Mozilla/Add-ons/WebExtensions/API/declarativeNetRequest/onRuleMatchedDebug_event",
```
Actually, this should be the other way around!  The `on` prefix should be removed but the `_event` suffix should stay.

jdatapple replied:
Addressed in 1ca88f3.

jdatapple commented on webextensions/declarative_net_request/rules.json:
```diff
@@ -0,0 +1,590 @@
+                }
+              }
+            },
+            "allowallrequests": {
```
Addressed in 1ca88f3.

jdatapple commented on webextensions/declarative_net_request/rules.json:
```diff
@@ -0,0 +1,590 @@
+{
+  "webextensions": {
+    "declarative_net_request": {
```
Addressed in 1ca88f3.

jdatapple commented on webextensions/declarative_net_request/rules.json:
```diff
@@ -0,0 +1,590 @@
+    "declarative_net_request": {
+      "rules": {
+        "__compat": {
+          "mdn_url": "https://developer.mozilla.org/docs/Mozilla/Add-ons/WebExtensions/manifest.json/declarative_net_request/rules",
```
Addressed in 1ca88f3.

jdatapple commented:
> Here's some feedback on the structure. It probably makes sense to do a full review once the version compatibility info is added + MDN docs are ready for review too.

I believe [1ca88f3](https://github.com/mdn/browser-compat-data/pull/14942/commits/1ca88f3539a0aa128de93fda81ddedf733d80899) addresses your feedback. Is this ready enough for me to start collecting browser data?

jdatapple commented:
@Rob--W or @queengooborg any other feedback on this so that we can continue to make progress?

Rob--W reviewed (changes requested):
> Is this ready enough for me to start collecting browser data?

By this, are you referring to `version_added`? With the changes below, the answer is yes.

You can skip Firefox for now, DNR is still under development there and there is no version that supports it yet.

queengooborg commented:
I happened to notice that every single feature here has no implementation yet.  We've actually implemented a policy _not_ to include features with no support in any browser in BCD to reduce clutter.  Are these features going to be implemented in Safari or other browsers?  Also, is this PR still planned?

Rob--W commented:
> I happened to notice that every single feature here has no implementation yet.  We've actually implemented a policy _not_ to include features with no support in any browser in BCD to reduce clutter.  Are these features going to be implemented in Safari or other browsers?  Also, is this PR still planned?

What lead to the conclusion that these features have no implementation?

There is an implementation in Chrome and Safari. And at the rate that the PR progresses, there may even be an implementation in Firefox when the PR+docs completes.

queengooborg commented:
> What lead to the conclusion that these features have no implementation?

In this PR, every feature added is set to `false` for all browsers.  I realize that there's been comments that clarify there's support; my apologies!

Since this PR has been sitting around for a long time, though, would you be down to create a new PR to supersede this one, @Rob--W?

Rob--W commented:
> Since this PR has been sitting around for a long time, though, would you be down to create a new PR to supersede this one, @Rob--W?

Despite the lack of visible progress on the PR, the PR has not been forgotten. Once the supported features have been clarified, and documentation for MDN is ready, I would expect more progress here.

If somehow the PR hasn't finished when Firefox's implementation ships, then I may consider taking over. But for now I think the PR is in good hands.

Rob--W commented:
This PR can be closed, because it has been superseded by #18403.
